src/coreclr/vm/interoplibinterface_comwrappers.cpp

```diff
@@ -672,1402 +48,682 @@ namespace
+    bool GetGlobalPeggingState() noexcept
     {
-        GCX_PREEMP();
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_ANY;
+        }
+        CONTRACTL_END;
 
-        HRESULT hr = externalComObject->QueryInterface(IID_IUnknown, &identity);
-        _ASSERTE(hr == S_OK);
+        return (VolatileLoad(&g_isGlobalPeggingOn) != FALSE);
     }
 
-    // Switch to Cooperative mode since object references
-    // are being manipulated.
+    bool IsObjectPromoted(_In_ InteropLib::OBJECTHANDLE handle) noexcept
     {
-        GCX_COOP();
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_COOPERATIVE;
+        }
+        CONTRACTL_END;
 
-        // TrackerObject support and unwrapping matches the built-in semantics that the global marshalling scenario mimics.
-        int flags = CreateObjectFlags::CreateObjectFlags_TrackerObject | CreateObjectFlags::CreateObjectFlags_Unwrap;
-        if ((objFromComIPFlags & ObjFromComIP::UNIQUE_OBJECT) != 0)
-            flags |= CreateObjectFlags::CreateObjectFlags_UniqueInstance;
-
-        // Passing NULL as the ComWrappers implementation indicates using the globally registered instance
-        return TryGetOrCreateObjectForComInstanceInternal(
-            NULL /*comWrappersImpl*/,
-            g_marshallingGlobalInstanceId,
-            identity,
-            NULL,
-            (CreateObjectFlags)flags,
-            ComWrappersScenario::MarshallingGlobalInstance,
-            NULL /*wrapperMaybe*/,
-            objRef);
-    }
-}
+        if (handle == nullptr)
+            return false;
 
-extern "C" void QCALLTYPE ComWrappers_SetGlobalInstanceRegisteredForTrackerSupport(INT64 id)
-{
-    QCALL_CONTRACT_NO_GC_TRANSITION;
+        ::OBJECTHANDLE objectHandle = static_cast<::OBJECTHANDLE>(handle);
 
-    _ASSERTE(g_trackerSupportGlobalInstanceId == ComWrappersNative::InvalidWrapperId && id != ComWrappersNative::InvalidWrapperId);
-    g_trackerSupportGlobalInstanceId = id;
-}
+        OBJECTREF obj = ObjectFromHandle(objectHandle);
 
-bool GlobalComWrappersForTrackerSupport::IsRegisteredInstance(INT64 id)
-{
-    return g_trackerSupportGlobalInstanceId != ComWrappersNative::InvalidWrapperId
-        && g_trackerSupportGlobalInstanceId == id;
-}
+        if (obj == nullptr)
+            return false;
 
-bool GlobalComWrappersForTrackerSupport::TryGetOrCreateComInterfaceForObject(
-    _In_ OBJECTREF instance,
-    _Outptr_ void** wrapperRaw)
-{
-    CONTRACTL
-    {
-        THROWS;
-        MODE_COOPERATIVE;
+        return GCHeapUtilities::GetGCHeap()->IsPromoted(OBJECTREFToObject(obj));
     }
-    CONTRACTL_END;
 
-    if (g_trackerSupportGlobalInstanceId == ComWrappersNative::InvalidWrapperId)
-        return false;
+    void SetGlobalPeggingState(_In_ bool state) noexcept
+    {
+        CONTRACTL
+        {
+            NOTHROW;
+            GC_NOTRIGGER;
+            MODE_ANY;
```

Comment:
Why can this be any?

Reply:
This is `ANY` in main today, I didn't change it.